ledger/complete/wal/checkpoint_v6_reader.go

```diff
@@ -326,37 +326,43 @@ func readNodeCount(f *os.File) (uint64, uint16, error) {
 	return decodeFooter(footer)
 }
 
-func computeTotalSubTrieNodeCount(groups [][]*node.Node) int {
+func computeTotalSubTrieNodeCount(groups [][]*node.Node) uint64 {
 	total := 0
 	for _, group := range groups {
-		total += (len(group) - 1) // the first item in group is <nil>
+		total += len(group)
 	}
-	return total
+	return uint64(total)
 }
 
-// each group in subtrieNodes and topLevelNodes start with `nil`
-func getNodeByIndex(subtrieNodes [][]*node.Node, topLevelNodes []*node.Node, index uint64) (*node.Node, error) {
+// get a node by node index.
+// Note: node index start from 1.
+// subtries contains subtrie node groups. subtries[i][0] is NOT nil.
+// topLevelNodes contains top level nodes. topLevelNodes[0] is nil.
+// any error returned are exceptions
+func getNodeByIndex(subtrieNodes [][]*node.Node, totalSubTrieNodeCount uint64, topLevelNodes []*node.Node, index uint64) (*node.Node, error) {
 	if index == 0 {
-		// item at index 0 is nil
+		// item at index 0 is for nil
 		return nil, nil
 	}
-	offset := index - 1 // index.> 0, won't underflow
+
+	if index > totalSubTrieNodeCount {
+		return getTopNodeByIndex(totalSubTrieNodeCount, topLevelNodes, index)
+	}
+
+	offset := index - 1 // index > 0, won't underflow
 	for _, subtries := range subtrieNodes {
-		if len(subtries) < 1 {
-			return nil, fmt.Errorf("subtries should have at least 1 item")
-		}
-		if subtries[0] != nil {
-			return nil, fmt.Errorf("subtrie[0] %v isn't nil", subtries[0])
+		if int(offset) < len(subtries) {
+			return subtries[offset], nil
 		}
 
-		if offset < uint64(len(subtries)-1) {
-			// +1 because first item is always nil
-			return subtries[offset+1], nil
-		}
-		offset -= uint64(len(subtries) - 1)
+		offset -= uint64(len(subtries))
 	}
 
-	// TODO: move this check outside
+	return nil, fmt.Errorf("could not find node by index %v, totalSubTrieNodeCount %v", index, totalSubTrieNodeCount)
+}
+
+func getTopNodeByIndex(totalSubTrieNodeCount uint64, topLevelNodes []*node.Node, index uint64) (*node.Node, error) {
```

Comment:
`getTopNodeByIndex` can be removed because we can lookup top level nodes directly in `getNodeByIndex`.

I think this function isn't necessary because:
- checking if `topLevelNodes[0]` is nil isn't necessary because this function is internal and only used by `readTopLevelTries`.  In `readTopLevelTries`, `topLevelNodes[0]` is nil and documented.
-  `if nodePos >= uint64(len(topLevelNodes))` isn't necessary because caller [filters out invalid node index](https://github.com/onflow/flow-go/blob/821bffa3d75c87bdfe954abd888f451c9824a6d1/ledger/complete/wal/checkpoint_v6_reader.go#L281-L283)  
- remaining functionality of looking up node in `topLevelNodes` with adjusted index can be moved to the caller.  See comment in `getNodeByIndex` for modified code.

Reply:
I do want to keep these checks, a checkpoint file might get manipulated in a certain way to bypass the early checks. 

For instance, a malicious actor might change the top level node to include an index for non-existing node, they could still produce a correct checksum, node count, etc, but then it could cause an out of index panic.

Reply:
@zhangchiqing I agree with you that we need to check for malicious data.  That is why we use multiple checks when processing checkpoint file.

However, the checks here don't detect malicious data because:
- a check is done to internal data structure not affected by potentially malicious data, or
- more strict check is already done by the caller.

In the first case, `topLevelNodes` is a slice we create and guarantee that the first element is nil.  No malicious data would change that guarantee.  

Your concern applies to the second case.  But a stricter check is already done in the caller.  https://github.com/onflow/flow-go/blob/821bffa3d75c87bdfe954abd888f451c9824a6d1/ledger/complete/wal/checkpoint_v6_reader.go#L281-L283

Given this, I think these specific checks here won't actually detect any malicious data in practice.  

Thoughts?